src/Lecture4.hs

```diff
@@ -135,7 +141,37 @@ errors. We will simply return an optional result here.
 -}
 
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow r = case wordsBy ',' r of
+   [p, t, c] -> do
+      p' <- parseProduct p
+      t' <- parseTradeType $ trim t
+      c' <- parseCost $ trim c
+      pure $ Row p' t' c'
+   _ -> Nothing
```

Comment:
You have an excellent implementation with neat pattern matching 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow r = do
    [p', t, c] <- Just $ wordsBy ',' r
    guard $ p' /= ""
    t' <- readMaybe t
    c' <- readMaybe c
    guard $ c' >= 0
    pure $ Row p' t' c'
```

You don't need to trim spaces manually as `read` functions do this for you 👍🏻